calico_node/startup/startup.go

```diff
@@ -312,35 +341,75 @@ func validateIP(ip *net.IP) {
+		// known internal interfaces).
+		return autoDetectCIDRFirstFound(version)
+	} else if strings.HasPrefix(method, AUTOD_METHOD_INTERFACE) {
+		// Autodetect the IP from the specified interface.
+		ifStr := strings.TrimPrefix(method, AUTOD_METHOD_INTERFACE)
+		return autoDetectCIDRByInterface(ifStr, version)
+	} else if strings.HasPrefix(method, AUTOD_METHOD_CAN_REACH) {
+		// Autodetect the IP by connecting a UDP socket to a supplied address.
+		destStr := strings.TrimPrefix(method, AUTOD_METHOD_CAN_REACH)
+		return autoDetectCIDRByReach(destStr, version)
+	}
+
+	// The autodetection method is not recognised and is required.  Exit.
+	fatal("Invalid IP autodection method: %s", method)
+	terminate()
+	return nil
+}
+
+// autoDetectCIDRFirstFound auto-detects the first valid Network it finds across
+// all interfaces (excluding common known internal interface names).
+func autoDetectCIDRFirstFound(version int) *net.IP {
+
 	incl := []string{}
-	excl := []string{"^docker.*", "^cbr.*", "dummy.*",
+	excl := []string{
+		"docker.*", "cbr.*", "dummy.*",
 		"virbr.*", "lxcbr.*", "veth.*", "lo",
-		"cali.*", "tunl.*", "flannel.*"}
-
-	// At the moment, we don't support anything other than the default
-	// (blank) auto-detection method.
-	if detectionMethod != "" {
-		fatal("IP detection method is not supported: %s", detectionMethod)
-		terminate()
+		"cali.*", "tunl.*", "flannel.*",
 	}
 
 	iface, cidr, err := autodetection.FilteredEnumeration(incl, excl, version)
 	if err != nil {
-		message("Unable to auto-detect any valid IPv%d addresses: %s", version, err)
+		warning("Unable to auto-detect an IPv%d address: %s", version, err)
 		return nil
 	}
 
-	if cidr == nil {
-		message("Unable to auto-detect an IPv%d address", version)
+	message("Using autodetected IPv%d address on interface %s: %s", version, iface.Name, cidr.String())
+
+	return &net.IP{cidr.IP}
+}
+
+// autoDetectCIDRByInterface auto-detects the first valid Network on the interfaces
+// matching the supplied interface regex.
+func autoDetectCIDRByInterface(ifaceRegex string, version int) *net.IP {
+	iface, cidr, err := autodetection.FilteredEnumeration([]string{ifaceRegex}, nil, version)
+	if err != nil {
+		warning("Unable to auto-detect an IPv%d address using interface regex %s: %s", version, ifaceRegex, err)
 		return nil
 	}
 
-	message("Using autodetected IPv%d address on interface %s: %s", version, iface.Name, cidr.String())
+	message("Using autodetected IPv%d address %s on matching interface %s", version, cidr.String(), iface.Name)
 
 	return &net.IP{cidr.IP}
 }
 
+// autoDetectCIDRByReach auto-detects the IP and Network by setting up a UDP
+// connection to a "reach" address.
+func autoDetectCIDRByReach(dest string, version int) *net.IP {
+	if cidr, err := autodetection.ReachDestination(dest, version); err != nil {
+		warning("Unable to auto-detect IPv%d address by connecting to %s: %s", version, dest, err)
+		return nil
+	} else {
```

Comment:
Not really.

Sometimes it makes the code look a little more "symmetric".  it'll compile to the same thing I would hope.